tests/unit/linalg/test_mma.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "unit_tests.hpp"
+#include "mfem.hpp"
+
+using namespace mfem;
+
+real_t obj0(mfem::Vector& x)
+{
+   const int n=x.Size();
+   real_t rez=0.0;
+   for (int i=0; i<n; i++)
+   {
+      rez=rez+x[i]*x[i];
+   }
+
+#ifdef MFEM_USE_MPI
+   real_t grez;
+   MPI_Allreduce(&rez, &grez, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+   rez = grez;
+#endif
+
+   return rez;
+}
+
+real_t dobj0(mfem::Vector& x, mfem::Vector& dx)
+{
+   const int n=x.Size();
+   real_t rez=0.0;
+   for (int i=0; i<n; i++)
+   {
+      rez=rez+x[i]*x[i];
+      dx[i]=2.0*x[i];
+   }
+#ifdef MFEM_USE_MPI
+   real_t grez;
+   MPI_Allreduce(&rez, &grez, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
```

Comment:
I think these all reductions on `real_t` should use `MPITypeMap<real_t>::mpi_type` instead of `MPI_DOUBLE` like [here](https://github.com/mfem/mfem/blob/master/linalg/hypre.cpp#L485).

Reply:
good point! changed!